src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs

```diff
@@ -114,6 +119,15 @@ internal void AddEntries(TreeRouteBuilder builder, ActionDescriptorCollection ac
                 }
             }
 
+            // Trim RouteInfo instances that do not partipate in inbound-routing.
+            for (var i = routeInfos.Count - 1; i >= 0; i--)
```

Comment:
What would you think about merging this into `GroupRouteInfos` and renaming that to something that makes it clear it's for inbound routes?

 Modifying this collection without a defensive copy seems like a trap for future generations.